File: 06_Algorithm/leetcode_814.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_814.py
# Author            : Andy
# Date              : 2022.09.19
# Last Modified Date: 2022.09.19
# Last Modified By  : Andy


class Solution:

    def pruneTree(self, root):

        def dfs(node):
            if node is None:
                return

            node.left = dfs(node.left)
            node.right = dfs(node.right)

            if node.left is None and node.right is None:
                # 有数字1的节点返回，如果没有则消除
                return node if node.val == 1 else None

            # 非叶子节点
            return node

        return dfs(root)

File: 06_Algorithm/leetcode_814.go
/**
 * File              : leetcode_814.go
 * Author            : Andy
 * Date              : 2022.09.19
 * Last Modified Date: 2022.09.19
 * Last Modified By  : Andy
 */

package pruneTree

func dfs(node *TreeNode) *TreeNode {
	if node == nil {
		return nil
	}

	Left := dfs(node.Left)
	node.Left = Left
	Right := dfs(node.Right)
	node.Right = Right

	if node.Left == nil && node.Right == nil {
		if node.Val == 1 {
			return node
		} else {
			return nil
		}
	}

	return node
}
func pruneTree(root *TreeNode) *TreeNode {

	return dfs(root)
}
